discord/voice_client.py

```diff
@@ -293,23 +393,180 @@ def is_connected(self):
+
+        Note
+        -----------
+        This function, although a co-routine, is not blocking.
+
+        Note
+        -----------
+        The returned data is stereo 48kHz audio in raw PCM. You will have to
+        add your own headers if you want to save it as a `.wav` file.
+        """
+
+        # Convert buffer_size from seconds to packets.
+        buffer_size = int(round((buffer_size * 1000) / 20))
+
+        SAMPLE_RATE = 48000
+        SAMPLE_SIZE = 2
+
+        last_packet = 0
+        user_id = user.id
+        voice_buffer = []
+        last_speaking = 0
+
+        def listener(vc, vp):
+            nonlocal voice_buffer, user_id, last_packet
+            if user_id == (None if vp.user is None else vp.user.id):
+                last_packet = vp.timestamp
+                heapq.heappush(voice_buffer, (vp.timestamp, vp))
+
+                if len(voice_buffer) > buffer_size:
+                    packet = heapq.heappop(voice_buffer)[1]
+
+                    if last_packet == 0:
+                        delta = 0
+                    else:
+                        delta = (packet.timestamp - last_packet) - 960
+
+                    if delta > 0:  # Ignore skipped packets
+                        data = bytearray([0] * delta * 4)
+                        data += packet.pcm
+
+                        file_object.write(data)
+
+                        last_packet = packet.timestamp
+
+        def speaking_listener(user, speaking):
+            nonlocal last_speaking, user_id
+            if user.id == user_id:
+                if speaking:
+                    if last_speaking != 0:
+                        delta = time.time() - last_speaking
+                        padding = bytearray([0] * int(delta * SAMPLE_RATE) * SAMPLE_SIZE)
+                        file_object.write(padding)
+
+                        last_speaking = 0
+                else:
+                    last_speaking = time.time()
+
+        self._pcm_listeners.append(listener)
+        self._speaking_listeners.append(speaking_listener)
+
+    @asyncio.coroutine
+    def _voice_receive_loop(self):
+        """
+        Listener loop.
+        This loop will receive data from `self.socket` then it will create a
+        `VoicePckaet` object that handles the decryption and decoding then
+        dispatch two client events so the user can tap into the data.
+        """
+
+        try:
+            while self.is_connected():
+                packet = yield from self.loop.sock_recv(self.socket, 65536)
+                log.debug('Received Voice Packet of {} bytes'.format(len(packet)))
+
+                vp = VoicePacket.unpack(packet, self)
+                log.debug('Decoded Voice Packet: {}'.format(vp))
+
+                if vp.ssrc not in self.decoders:
+                    self.decoders[vp.ssrc] = opus.Decoder()
+
+                # Dispatch event
+                self._state.dispatch('receive_opus', self, vp)
+
+                # TODO: re-order packets
```

Comment:
you're already doing this to some effect in the listener. you should just move this decode logic into there.

Reply:
The listener and the events are separate. The listens are for when a user uses one of the pre-built functions to receive audio. The events send the packets as they arrive. This means the listeners have the luxury of being able to buffer the entire data before reordering it, whereas I will need a genuine jitter buffer for the events.